lib/stdlib/src/binary.erl

```diff
@@ -934,6 +935,31 @@ unhex(X) ->
              no, 10, 11, 12, 13, 14, 15, no, no, no, no, no, no, no, no, no  %96
             }).
 
+-doc """
+Joins a list of binaries together by a specified `Separator`.
+
+Equivalent to `iolist_to_binary(lists:join(Separator, Binaries))` with `binary:join/2` being the faster of the two.
+
+_Example:_
+
+```erlang
+1> binary:join([<<"a">>, <<"b">>, <<"c">>], <<", ">>).
+<<"a, b, c">>
+```
+""".
+-doc(#{since => <<"OTP 27.0">>}).
+-spec join([binary()], binary()) -> binary().
+join([], _Separator) -> <<>>;
+join([H], _Separator) -> H;
```

Comment:
```suggestion
join([H], _Separator) when is_binary(H) -> H;
```

Reply:
I noticed the same bug 😄 Had a test case that failed on me 😄

Reply:
Now that I think about it, you should also check the `Separator` in the empty and single-element list clauses :sweat_smile:

Reply:
I thought about it too but technically speaking it doesn't change the outcome. So while a `badarg` should be raised from that perspective, the outcome being `<<>> | H` wouldn't make use of the `Separator` so technically it's an unnecessary guard with little gain IMHO. 

Happy to include it though if more agree 👍

Reply:
Actually it might make a small difference in performance to validate the type of separator before the main loop - the compiler will likely omit those typechecks later on each loop iteration